mpqp/execution/providers/azure.py

```diff
@@ -0,0 +1,202 @@
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING, Optional
+
+import numpy as np
+
+if TYPE_CHECKING:
+    from qiskit.result import Result as QiskitResult
+    from azure.quantum.target.microsoft.result import MicrosoftEstimatorResult
+
+
+from typeguard import typechecked
+
+from mpqp.core.circuit import QCircuit
+from mpqp.core.instruction.measurement import BasisMeasure
+from mpqp.core.languages import Language
+from mpqp.execution.connection.azure_connection import (
+    get_azure_provider,
+    get_jobs_by_id,
+)
+from mpqp.execution.devices import AZUREDevice
+from mpqp.execution.job import Job, JobStatus, JobType
+from mpqp.execution.result import Result, Sample, StateVector
+
+
+@typechecked
+def run_azure(job: Job) -> Result:
+    """Executes the job on the right AZURE device precised in the job in
+    parameter.
+
+    Args:
+        job: Job to be executed.
+
+    Returns:
+        The result of the job.
+
+    Note:
+        This function is not meant to be used directly, please use
+        :func:``run<mpqp.execution.runner.run>`` instead.
+    """
+    from qiskit import QuantumCircuit
+
+    qiskit_circuit = (
+        job.circuit.without_measurements().to_other_language(Language.QISKIT)
+        if (job.job_type == JobType.STATE_VECTOR)
+        else job.circuit.to_other_language(Language.QISKIT)
+    )
+    if TYPE_CHECKING:
+        assert isinstance(qiskit_circuit, QuantumCircuit)
+
+    qiskit_circuit = qiskit_circuit.reverse_bits()
+
+    backend_sim = get_azure_provider().get_backend(job.device.value)
+
+    assert isinstance(job.device, AZUREDevice)
+
+    # if job.job_type == JobType.STATE_VECTOR:
+    #     qiskit_circuit.save_statevector() pyright: ignore[reportAttributeAccessIssue]
+    #     job.status = JobStatus.RUNNING
+    #     job_sim = backend_sim.run(qiskit_circuit, shots=0)
+    #     result_sim = job_sim.result()
+    #     result = extract_result(result_sim, job, job.device)
+    if job.job_type == JobType.SAMPLE:
+        assert job.measure is not None
+        job.status = JobStatus.RUNNING
+        job_sim = backend_sim.run(qiskit_circuit, shots=job.measure.shots)
+        result_sim = job_sim.result()
+        result = extract_result(result_sim, job, job.device)
+    else:
+        raise ValueError(f"Job type {job.job_type} not handled.")
+
+    job.status = JobStatus.DONE
+    return result
+
+
+@typechecked
+def extract_result(
+    result: "MicrosoftEstimatorResult | QiskitResult",
+    job: Optional[Job],
+    device: AZUREDevice,
+) -> Result:
+    from qiskit.result import Result as QiskitResult
+
+    if isinstance(result, QiskitResult):
```

Comment:
can't we use the qiskit result extractor for this ?

Reply:
No, it is not exactly the same result, and it causes trouble.

Reply:
this is fucked up :') can you tell me more ?

Reply:
for azure:
`data = [
                Sample(
                    bin_str="".join(map(str, state)),
                    nb_qubits=job.circuit.nb_qubits,
                    count=int(count),
                    probability=(
                        job_data.get("probabilities").get(state)
                        if "probabilities" in job_data
                        else None
                    ),
                )
                for (state, count) in job_data.get("counts").items()
            ]
`
for qiskit:
`data = [
                    Sample(
                        bin_str=item,
                        count=counts[item],
                        nb_qubits=job.circuit.nb_qubits,
                        probability=(
                            job_data.get("probabilities").get(item)
                            if "probabilities" in job_data
                            else None
                        ),
                    )
                    for item in counts
                ]
`

for both cases is a QiskitResult, but they are not the same elements, so it's a mess to make just one code.